### ISSUE TITLE
Voxelwise correlation matrix calculation takes forever 

### ISSUE THREAD

**Issue by wizofe** [nilearn/nilearn#3206]
Using a 3mm isotropic fMRI for voxelwise whole brain correlation takes forever, while if I just extract the brain signal using a GM mask and use numpy functions is relatively fast (for the 3mm numpy takes about 5 minutes while nilearn was running for at least 1 hour before I stopped it).

Nilearn version: 0.9.0

### Expected behavior
I am trying to calculate a connectivity matrix for voxelwise brain correlation. I am using an MNI152 registered fMRI NIfTI with 3mm and 4mm isotropic resolution and a gray mask. 

### Actual behavior
For the 3mm isotropic the process takes forever and I don't even know how long it needs to complete, but I expect more than an hour on my machine. For the 4mm isotropic it does finish in a reasonable amount of time but still it isn't as fast as the `np.corrcoef`. 

I would like to see if it is possible to accelerate this process or if it is an issue on the calculation algorithm as it seems to work forever. Happy to get pointed to the right location to debug as well.

### Steps and code to reproduce bug

```python
from nilearn.maskers import NiftiMasker
from nilearn.connectome import ConnectivityMeasure

import nibabel as nib
import numpy as np

img = nib.load("brain_4mm.nii.gz")
mask = nib.load("mask_4mm.nii.gz")

masker = NiftiMasker(mask, memory='nilearn_cache', memory_level=1, verbose=0, smoothing_fwhm=None)
braints = masker.fit_transform(img)

cor_mes = ConnectivityMeasure(kind='correlation')
cor_mat = cor_mes.fit_transform([braints])[0]
```
but if I calculate the correlation of the brain time series using `np.corrcoef(braints.T)` it takes significantly less.

**Comment by bthirion** [comment]
But what you commute with `np.corrcoef(braints.T)` is not a connectivity model, but an across-time similarity measure, which may be interesting on its own, but is not a connectivity measure.

**Comment by bthirion** [comment]
I would heavily discourage computing a full voxel-by-voxel correlation matrix, which yields unmanageably large arrays to store low-rank information.

**Comment by bthirion** [comment]
In any case, I'm not sure I identifyu what you call a bug: is it that this takes ages to compute on a 3mm-image ?

**Comment by wizofe** [comment]
Thank you @bthirion. Specifically what I call a bug is that it takes extremely long time to compute a 3mm isotropic vs a 4mm isotropic fMRI. 

I may have misunderstood the difference between a connectivity measure then. What I am looking for is to create a connectivity matrix using the Pearson correlation of voxelwise whole brain similarity, in order to calculate some graph metrics (such as degree centrality).

Would you point me on the implementation of this and let me know if there's a way to directly implement it in numpy so I could benchmark it? Thanks!

**Comment by emdupre** [comment]
Hi @wizofe,

I agree with @bthirion that it's hard to call this behavior a bug. I'll go ahead and re-label this issue as an `Enhancement`, but we can re-label later if it the discussion clarifies something in the current code base that should be fixed rather than optionally improved.

To your question of computing a voxelwise connectivity matrix, though: this is not something Nilearn explicitly supports (yet) since, as mentioned in-thread, it yields quite large arrays and therefore requires unique optimizations compared to other more connectivity measures using e.g. ROIs or seeds. 

There are other library implementations, though, that specifically target this use case; for example, the [Full Correlation Matrix Analysis (FCMA) of Brainiak](https://brainiak.org/tutorials/09-fcma/). You could check their code base for the full implementation ! If I recall correctly, it's likely a mix of Python and C. Hope that helps !

**Comment by alexisthual** [comment]
@wizofe to add to what @emdupre is saying, if you need to do computations using large matrices whose mathematical expression is simple (such as a correlation matrix), you could try using [pykeops](https://github.com/getkeops/keops).

**Comment by emdupre** [comment]
I'll close this now, but if you see a specific path forward @wizofe, please open a new issue ! We're also very happy to discuss more general usage questions on [Neurostars](https://neurostars.org).
